### PR TITLE
Allow submodules

### PR DESCRIPTION
**Pull request by dfalbel**
This PR is an attempt to allow submodules as suggested by @andy-thomason here https://github.com/extendr/extendr/issues/140#issuecomment-764071735

The missing part would be to allow the submodule to be exported with something like this, ie. without the `mod` declaration.

```
extendr_module! {
  fn my_function;
  impl my_impl;
}
```
I couldn't find a way to do this because the `modulename` is used to generate the `get_{modname}_metadata` function that is then used to fuse the main module metadata with submodules.

Thus the `init__*` functions are still generated, but not used, because the metadata is included in the main module when using:

```
extendr_module! {
  mod dll_name;
  use my_submodule;
}
```

I'd appreciate some guidance on how to solve this issue.

### CODE REVIEW

**Comment by clauswilke** [comment]
See also #151, which may be related.

Would it be possible to give submodules their own module name and then somehow add them to the main module with something like `use my_submodule;`? 

In any case, all functions to be registered with R need to be collected and registered at once, we cannot separately call the current registration functions for submodules because they will overwrite previous registrations.

**Comment by dfalbel** [comment]
Yeah, this is what this PR is doing. In `submodule.rs` we need to declare:

```
extendr_module! {
    mod submodule;
    fn hello_submodule;
    impl MySubmoduleClass;
}
```

And in `lib.rs` we do:

```
use extendr_api::prelude::*;
mod submodule;
use submodule::*;

extendr_module! {
    mod extendrtests;
    use submodule;
}
```

With this the `R_init_submodule_extendr` is still created, but it's never used because all registration happens in `R_init_extendrtests_extendr`.

**Comment by clauswilke** [comment]
I guess I'm not sure I fully understand the issue then. Is it just aesthetics, that there is a second init routine that is generated but not used, or is there a problem with functionality (functions that should be registered but aren't)?

I think all modules and submodules will always need a name. So this will never work I think:
```
extendr_module! {
  fn my_function;
  impl my_impl;
}
```

**Comment by dfalbel** [comment]
AFAICT it's just aesthetic. All functionality seems to work fine, it's just that we register something that we don't use...

In theory, it would be possible to have 'unnamed' modules and generate `submodule::get_metadata` instead of `submodule::get_submodule_metadata`, ie. let the namespace prefix to disambiguate, but it might be better to keep the names anyway.

**Comment by clauswilke** [comment]
Yeah, looks good to me. I'm not so worried about extra registration functions. Who knows, maybe somebody can do some nifty things with it. (You could make one Rust library behave in two different ways in R, depending on which registration function you call. Not sure if that's in any way useful, but maybe it is.)

Are the wrapper generator functions getting registered, i.e., `wrap__make_<submodule>_wrappers()`? Can I generate wrappers just for a submodule? (This would be great for testing, where we could have a separate small submodule where we test that the wrappers come out just as expected, and that would never change even as we add more integration tests otherwise.)

File: extendr-macros/src/lib.rs [hunk 1]
```diff
@@ -710,6 +719,9 @@ pub fn extendr_module(item: TokenStream) -> TokenStream {
                 hidden: true,
             });
 
+            #( functions.extend(#usemetanames().functions); )*
+            #( impls.extend(#usemetanames().impls); )*
```
Comment (clauswilke) [inline extendr-macros/src/lib.rs:723]: Maybe move these two lines up to where the module functions and implementations are added? (I.e., above the metadata access function.)

**Review by clauswilke — APPROVED** [review]
While you're at it, could you add a comment to these two lines:

```
#( #fnmetanames(&mut functions); )*
#( #implmetanames(&mut impls); )*
```
Something like `// add function metadata` or similar? It took me a while to figure out what these two functions do, so adding a comment may make this code more maintainable in the future.

**Comment by dfalbel** [comment]
Thanks @clauswilke  for the review!

> Yeah, looks good to me. I'm not so worried about extra registration functions. Who knows, maybe somebody can do some nifty things with it. (You could make one Rust library behave in two different ways in R, depending on which registration function you call. Not sure if that's in any way useful, but maybe it is.)

Makes sense to me! I think having the same code path is more valuable here and simplifies the implementation.

> Are the wrapper generator functions getting registered, i.e., wrap__make_<submodule>_wrappers()? Can I generate wrappers just for a submodule? (This would be great for testing, where we could have a separate small submodule where we test that the wrappers come out just as expected, and that would never change even as we add more integration tests otherwise.)

Yeah, it seems that `wrap__make_<submodule>_wrapers` is registered:

``` r
library(extendrtests)
.Call(extendrtests:::wrap__make_submodule_wrappers, TRUE, "extendrtests")
#> [1] "# Generated by extendr: Do not edit by hand\n#\n# This file was created with the following call:\n#   .Call(\"wrap__make_submodule_wrappers\", use_symbols = TRUE, package_name = \"extendrtests\")\n\n#' @docType package\n#' @usage NULL\n#' @useDynLib extendrtests, .registration = TRUE\nNULL\n\n#' Return string `\"Hello world!\"` to R.\n#' @export\nhello_submodule <- function() .Call(wrap__hello_submodule)\n\n#' Class for testing (exported)\n#' @examples\n#' x <- MySubmoduleClass$new()\n#' x$a()\n#' x$set_a(10)\n#' x$a()\n#' @export\nMySubmoduleClass <- new.env(parent = emptyenv())\n\nMySubmoduleClass$new <- function() .Call(wrap__MySubmoduleClass__new)\n\nMySubmoduleClass$set_a <- function(x) invisible(.Call(wrap__MySubmoduleClass__set_a, self, x))\n\nMySubmoduleClass$a <- function() .Call(wrap__MySubmoduleClass__a, self)\n\nMySubmoduleClass$me <- function() .Call(wrap__MySubmoduleClass__me, self)\n\n#' @rdname MySubmoduleClass\n#' @usage NULL\n#' @export\n`$.MySubmoduleClass` <- function (self, name) { func <- MySubmoduleClass[[name]]; environment(func) <- environment(); func }\n\n"
```

> Maybe move these two lines up to where the module functions and implementations are added? (I.e., above the metadata access function.)

Makes sense! I'll move the code up and add comments! 

Thanks again!

**Comment by clauswilke** [comment]
There's a small code formatting problem now. Do you think it's ready to be merged once that is fixed?

**Comment by dfalbel** [comment]
Fixed the code formatting issue! I think this is ready to merge.

**Comment by clauswilke** [comment]
Thanks!

**Pull request merged**